fuse/managers/pipeline.py

```diff
@@ -0,0 +1,277 @@
+        custom parameters for infer_func.
+    :param eval_params: Dictionary that can contain any number of additional
+        custom parameters for eval_func.
+    :param sample_ids_per_fold: May contain a sequence of array pairs denoting sample ids
+        for pre-defined train/validation splits. If this parameter is set to `None`,
+        the splits are decided at random.
+    """
+
+    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"  # required for pytorch deterministic mode
+    multiprocessing.set_start_method("spawn")
+    if num_gpus_total == 0 or num_gpus_per_split == 0:
+        if train_params is not None and "manager.train_params" in train_params:
+            train_params["manager.train_params"]["device"] = "cpu"
+
+    # set debug mode:
+    mode = "default"  # Options: 'default', 'debug'. See details in FuseDebug
+    _ = FuseDebug(mode)
+
+    available_gpu_ids = FuseUtilsGPU.get_available_gpu_ids()
+    if num_gpus_total < len(available_gpu_ids):
+        available_gpu_ids = available_gpu_ids[0:num_gpus_total]
+    # group gpus into chunks of size params['common']['num_gpus_per_split']
+    gpu_resources = [
+        available_gpu_ids[i : i + num_gpus_per_split] for i in range(0, len(available_gpu_ids), num_gpus_per_split)
+    ]
+
+    # create a queue of gpu chunks (resources)
+    q_resources = Queue()
+    for r in gpu_resources:
+        q_resources.put(r)
+
+    if sample_ids_per_fold is None:
+        dataset, test_dataset = dataset_func(cache_dir=paths["cache_dir"], **dataset_params)
+        # the split decision should be the same regardless of repetition index
+        kfold = KFold(n_splits=num_folds, shuffle=True, random_state=1234)
+        sample_ids_per_fold = [item for item in kfold.split(dataset)]
+    else:
+        assert num_folds == len(sample_ids_per_fold)
+
+    for rep_index in range(num_repetitions):
+        # run training, inference and evaluation on all cross validation folds in parallel
+        # using the available gpu resources:
+        runner = partial(runner_wrapper, q_resources, rep_index, [train_wrapper, infer_wrapper, eval_wrapper])
+
+        # create process per fold
+        processes = [
+            Process(target=runner, args=(ids, cv_index, rep_index, dataset_func, dataset_params, paths, [train_func, infer_func, eval_func], [train_params, infer_params, eval_params]))
+            for (ids, cv_index) in zip(sample_ids_per_fold, range(num_folds))
+        ][0:num_folds_used]
+        for p in processes:
+            p.start()
+
+        for p in processes:
+            p.join()
+            p.close()
+
+        # infer and eval each split's model on test set:
+        runner = partial(runner_wrapper, q_resources, rep_index, [infer_wrapper, eval_wrapper])
+        # create process per fold
+        processes = [
+            Process(target=runner, args=(None, cv_index, rep_index, dataset_func, dataset_params, paths, [infer_func, eval_func], [infer_params, eval_params]))
```

Comment:
The ids=None seems relevant for mnist and not necessarily for the others. you might want to give the user an option to specify test ids separately (and he can set it to None if required)

Reply:
currently the design assumption is that the user needs to implement a `create_dataset` function that should return two datasets: either train+val (with given sample id pair) or dev+test (without sample id pair, because it's just ALL the available data for dev, and ALL the available data for test)
so under this assumption, it makes sense to pass None during test
do you have a better suggestion?

how would you define the requirements from the `create_dataset` function?